### DISCUSSION TITLE
Neovim and Hybris

### DISCUSSION THREAD

**Discussion by plemanski** [mfussenegger/nvim-jdtls#530] [Q&A] [answered]
Hello,

I work on a combination of microservices and Hybris (SAP Commerce). If anyone is familiar with Hybris, you will be familiar with the project structure. My current job is my first time working with Hybris. I use Neovim for 99% of my development. However, I have not been able to find a way to get JDTLS to recognize dependencies in a hybris project. This has forced me to use Intellij when working with hybris. I would much rather not pay for IntelliJ for when I need to hop into Hybris. 

Has anyone successfully used neovim with jdtls in a hybris project?

**Comment by diegoortizmatajira** [comment]
@plemanski Did you have any progress since your question?

**Comment by manuelestebanpr** [comment]
Hi @plemanski . I mostly use AI, but you can get it to work if you properly scan the platform and your custom extension folders. Make sure that in your custom extensions, you use the extensioninfo.xml to import dependencies as well. I don't think there is a need to import modules if they aren't called from the custom extensions or elsewhere on the platform, but checking the extension info will help you a lot.

Also, I use LemMinX for XML and try to sync it with the *.xsd files.

**Reply by diegoortizmatajira** [reply, marked as answer]
Hi guys @plemanski @manuelestebanpr 

@manuelestebanpr I looked at your repo, and it gave me good ideas... Thanks a lot!

I have integrated them into my personal Neovim setup: https://github.com/diegoortizmatajira/LYRD-lua

My approach consists of:
- Automatic extension scanning: detects the Hybris installation (HYBRIS_HOME), reads localextensions.xml for active extensions, resolves transitive requires-extension dependencies, and indexes each extension's source paths (from .classpath, with heuristic fallbacks for src, gensrc, web/commonwebsrc, backoffice/src, hmc/src, etc.) and jars (extension lib/bin, platform lib, bootstrap/bin, tomcat/lib).
- Scan results are cached per project and reused on startup; a full rescan or cache reload is available via custom commands: LYRDJavaHybrisImportSolution, LYRDJavaHybrisReloadSolution, and LYRDJavaHybrisConfigureSolution.
- Wires the scanned classpath and source paths into JDT.LS (java.project.referencedLibraries, sourcePaths, java.import.exclusions) so Hybris extensions are treated as one workspace instead of per-extension Eclipse projects.
- I have added an UI to check/uncheck extensions to be loaded into the workspace.
- I have overseer plugin configured with additional tasks to run: server start, stop, debug, ant build, etc.

Key parts: 
- https://github.com/diegoortizmatajira/LYRD-lua/blob/main/layers/lang/java.lua
- https://github.com/diegoortizmatajira/LYRD-lua/blob/main/layers/lang/java-hybris.lua
- https://github.com/diegoortizmatajira/LYRD-lua/blob/main/runtime/lsp/jdtls.lua
- https://github.com/diegoortizmatajira/LYRD-lua/tree/main/shared/hybris
- https://github.com/diegoortizmatajira/LYRD-lua/blob/main/shared/ui/hybris_solution.lua
- https://github.com/diegoortizmatajira/LYRD-lua/blob/main/shared/overseer/hybris_tasks.lua

<img width="1917" height="1138" alt="image" src="https://github.com/user-attachments/assets/0322d2e0-2338-444d-bf5f-071ea32e22e1" />

<img width="1360" height="1056" alt="image" src="https://github.com/user-attachments/assets/8a226dfc-bd15-4da3-bb5d-b7e4058b0d3b" />

<img width="1535" height="962" alt="image" src="https://github.com/user-attachments/assets/5c083809-cbb5-4139-aa63-4295ce22091e" />
